File: hackerrank/problem solving/algorithms/2 - implementation/32 - sherlock and squares/sherlock_and_squares.go
// https://www.hackerrank.com/challenges/sherlock-and-squares/problem?isFullScreen=true

package main

import (
    "fmt"
    "math"
)

func main() {
    var n int
    fmt.Scan(&n)
    for range n {
        var a, b int
        _, _ = fmt.Scan(&a, &b)
        fmt.Println(squares(a, b))
    }
}

// T: O(1)
// S: O(1) extra space
func squares(startNum, endNum int) int {
    maxSquare := int(math.Floor(math.Sqrt(float64(endNum))))
    minSquare := int(math.Ceil(math.Sqrt(float64(startNum))))
    return maxSquare - minSquare + 1
}

File: hackerrank/problem solving/algorithms/2 - implementation/32 - sherlock and squares/sherlock_and_squares.cs
// https://www.hackerrank.com/challenges/sherlock-and-squares/problem?isFullScreen=true

int n = ReadNums()[0];
for (int i = 0; i < n; i++)
{
    var range = ReadNums();
    Console.WriteLine(Squares(range[0], range[1]));
}

int[] ReadNums() => Console.ReadLine()!.Split().Select(int.Parse).ToArray();

// T: O(1)
// S: O(1) extra space
int Squares(int startNum, int endNum)
{
    int maxSquare = (int)Math.Floor(Math.Sqrt(endNum));
    int minSquare = (int)Math.Ceiling(Math.Sqrt(startNum));
    return maxSquare - minSquare + 1;
}
